addons/crm/models/crm_lead.py

```diff
@@ -161,8 +161,15 @@ class CrmLead(models.Model):
     date_conversion = fields.Datetime('Conversion Date', readonly=True)
     date_deadline = fields.Date('Expected Closing', help="Estimate of the date on which the opportunity will be won.")
     # Customer / contact
+    company_partner_id = fields.Many2one('res.partner', string='Company Partner', compute="_compute_partners", inverse="_inverse_partners", domain="[('is_company', '=', True)]")
+    contact_partner_id = fields.Many2one('res.partner', string='Customer Partner', compute="_compute_partners", inverse="_inverse_partners",
+        domain="[('is_company', '!=', True)] + ([('id', 'child_of', company_partner_id)] if company_partner_id else [])"
+    )
+    company_has_children = fields.Boolean(compute='_compute_company_has_children')
+    contact_has_parent = fields.Boolean(compute='_compute_contact_has_parent')
     partner_id = fields.Many2one(
         'res.partner', string='Customer', check_company=True, index=True, tracking=10,
+        compute="_compute_partner_id", readonly=False, store=True,
```

Comment:
Seems weird to make it a compute (it is not really a compute), and also weird to have it computed based on fields available in a given view, that are computed / inverse. All this is a bit complicated at model level for just a helper in kanban view.

Can't we left main fields untouched, and just have interface fields with onchange ? I know, old school, but onchange are great when we need something only for UX / view purpose without impacting too much models.

Reply:
I agree in principle. It's needed for the edge case where you have a contact with no company, and you then set a company.

It's to do with update order and compute protection, as the fields are defined recursively.
Yes, it's very sucky.